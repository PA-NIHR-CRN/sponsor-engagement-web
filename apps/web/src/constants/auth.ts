/**
 * Identifer of our custom next-auth provider
 */
export const AUTH_PROVIDER_ID = 'oidc'

/**
 * Name of our custom next-auth provider
 */
export const AUTH_PROVIDER_NAME = 'OIDC'

/**
 * Auth type of our custom next-auth provider
 */
export const AUTH_PROVIDER_TYPE = 'oauth'

/**
 * Fallback session expiry to use when the NEXTAUTH_SESSION_EXPIRY env var is not provided
 * This is specifically for next-auth's JWT session expiration and not IDG's access token expiration
 */
export const AUTH_SESSION_EXPIRY_FALLBACK = 2592000 // 30 days

/**
 * Local roles associated with a user
 */
export enum Roles {
  SponsorContact = 1,
  ContactManager = 2,
}

export const PASSWORD_MIN_LENGTH = 12

export const ODP_ROLE = 'ODP_SponsorEngagementTool'
