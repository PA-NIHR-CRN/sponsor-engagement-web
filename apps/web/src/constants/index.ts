export * from './filters'
export * from './content'
export * from './identifiers'
export * from './auth'
