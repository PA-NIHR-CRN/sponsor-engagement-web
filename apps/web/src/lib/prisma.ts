import { PrismaClient } from '@prisma/client'

export const prismaClient = new PrismaClient()

// Getting the above via the database package is erroring with "PrismaClient is not a constructor"

// import { PrismaClient } from 'database'

// export const prismaClient = new PrismaClient()
