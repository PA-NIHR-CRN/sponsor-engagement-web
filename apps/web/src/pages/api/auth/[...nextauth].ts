import type { AuthOptions } from 'next-auth'
import NextAuth from 'next-auth'
import type { OAuthConfig, OAuthUserConfig } from 'next-auth/providers'
import { AUTH_PROVIDER_ID, AUTH_PROVIDER_NAME, AUTH_PROVIDER_TYPE } from '@/constants/auth'

interface Profile {
  id: string
  given_name: string
  family_name: string
  email: string
}

type ProviderOptions = OAuthUserConfig<'type'> & { wellKnown: string }

const Provider = ({ clientId, clientSecret, wellKnown }: ProviderOptions): OAuthConfig<Profile> => ({
  id: AUTH_PROVIDER_ID,
  name: AUTH_PROVIDER_NAME,
  type: AUTH_PROVIDER_TYPE,
  wellKnown,
  clientId,
  clientSecret,
  authorization: { params: { scope: 'openid email profile' } },
  idToken: true,
  checks: ['pkce', 'state'],
  profile(profile) {
    return {
      id: String(profile.id),
      firstName: profile.given_name,
      lastName: profile.family_name,
      emailAddress: profile.email,
    }
  },
})

export const authOptions: AuthOptions = {
  debug: true,
  providers: [
    Provider({
      clientId: process.env.AUTH_CLIENT_ID,
      clientSecret: process.env.AUTH_CLIENT_SECRET,
      wellKnown: process.env.AUTH_WELL_KNOWN_URL,
    }),
  ],
}
export default NextAuth(authOptions)
