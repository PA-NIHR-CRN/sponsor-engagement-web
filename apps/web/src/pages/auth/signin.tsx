import { signIn, useSession, getCsrfToken, getProviders } from 'next-auth/react'
import { useEffect } from 'react'
import { useRouter } from 'next/router'
import type { GetServerSidePropsContext, InferGetServerSidePropsType } from 'next'
import { Container } from '@nihr-ui/frontend'
import { AUTH_PROVIDER_ID } from '@/constants/auth'

/**
 * Our middleware.ts redirects to this sign in page for unauthenticated users.
 *
 * This signin page then automatically redirects to the OIDC sign in page or back to our application root depending on authentication status
 * next-auth does not support redirecting to the authentication provider directly from the middleware. (https://github.com/nextauthjs/next-auth/discussions/4511)
 */
export type SigninProps = InferGetServerSidePropsType<typeof getServerSideProps>

export default function Signin({ signinUrl, csrfToken, callbackUrl }: SigninProps) {
  const router = useRouter()
  const { status } = useSession()

  useEffect(() => {
    if (status === 'unauthenticated') {
      void signIn(AUTH_PROVIDER_ID)
    } else if (status === 'authenticated') {
      void router.push('/')
    }
  }, [status, router])

  // Fallback for non-Javascript
  return (
    <Container>
      <form action={signinUrl} method="POST">
        <input name="csrfToken" type="hidden" value={csrfToken} />
        <input name="callbackUrl" type="hidden" value={callbackUrl} />
        <button className="govuk-button" type="submit">
          Sign in
        </button>
      </form>
    </Container>
  )
}

export const getServerSideProps = async ({ req }: GetServerSidePropsContext) => {
  try {
    const providers = await getProviders()

    if (providers?.oidc) {
      const { signinUrl } = providers.oidc

      return {
        props: {
          signinUrl,
          csrfToken: await getCsrfToken({ req }),
          callbackUrl: `${process.env.NEXTAUTH_URL}/`,
        },
      }
    }

    throw new Error('Could not retrieve auth providers')
  } catch (error) {
    return {
      redirect: {
        destination: '/500',
      },
    }
  }
}
