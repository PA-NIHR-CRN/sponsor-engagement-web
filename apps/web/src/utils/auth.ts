import { Roles } from '../constants/auth'

export class AuthError extends Error {
  constructor(message: string) {
    super(message)
    this.name = this.constructor.name
  }
}

export const isContactManager = (roles: number[]) => roles.length === 1 && roles.includes(Roles.ContactManager)

export const isSponsorContact = (roles: number[]) => roles.length === 1 && roles.includes(Roles.SponsorContact)

export const isContactManagerAndSponsorContact = (roles: number[]) =>
  roles.includes(Roles.ContactManager) && roles.includes(Roles.SponsorContact)
