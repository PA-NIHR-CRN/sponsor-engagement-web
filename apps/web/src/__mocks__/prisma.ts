import type { PrismaClient } from 'database'
import type { DeepMockProxy } from 'jest-mock-extended'
import { mockDeep, mockReset } from 'jest-mock-extended'
import { prismaClient } from '../lib/prisma'

jest.mock('../lib/prisma', () => ({
  prismaClient: mockDeep<PrismaClient>(),
}))

beforeEach(() => mockReset(prismaMock))

export const prismaMock = prismaClient as DeepMockProxy<PrismaClient>
