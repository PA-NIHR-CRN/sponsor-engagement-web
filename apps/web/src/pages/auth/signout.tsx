import { getCsrfToken, getProviders, signOut } from 'next-auth/react'
import { useEffect } from 'react'
import type { GetServerSidePropsContext, InferGetServerSidePropsType } from 'next'
import { Container } from '@nihr-ui/frontend'
import { getServerSession } from 'next-auth/next'
import { logger } from '@nihr-ui/logger'
import { authOptions } from '../api/auth/[...nextauth]'
import { SIGN_IN_PAGE } from '../../constants/routes'
import { prismaClient } from '../../lib/prisma'

/**
 * This sign out page provides a progressively enhanced sign out experience
 *
 * For users w/ JavaScript it will automatically clear the session using next-auth's
 * built in `signOut` method (which is unfortunately client-side only).
 *
 * For users w/o JavaScript, a simple form is presented which makes a POST request to
 * next-auth's signout API handler.
 *
 * Both options will then redirect to IDG for the user to complete a federated logout
 */
export type SignoutProps = InferGetServerSidePropsType<typeof getServerSideProps>

export default function Signout({ csrfToken, isAuthenticated, callbackUrl }: SignoutProps) {
  useEffect(() => {
    const handleSignout = async () => {
      await signOut({ callbackUrl })
    }
    if (isAuthenticated) {
      handleSignout().catch(console.error)
    }
  }, [isAuthenticated, callbackUrl])

  // Fallback for non-Javascript
  return (
    <Container>
      <h2 className="govuk-heading-l">Signing out...</h2>
      <form action="/api/auth/signout" method="POST">
        <input name="callbackUrl" type="hidden" value={callbackUrl} />
        <input name="csrfToken" type="hidden" value={csrfToken} />
        <button className="govuk-link" type="submit">
          Click here if you are not redirected
        </button>
      </form>
    </Container>
  )
}

export const getServerSideProps = async (context: GetServerSidePropsContext) => {
  try {
    const session = await getServerSession(context.req, context.res, authOptions)

    const { id_token: idTokenHint } = await prismaClient.account.findFirstOrThrow({
      where: {
        userId: session?.user?.id,
      },
      select: {
        id_token: true,
      },
    })

    // If there's no session, redirect to sign in
    if (!session) {
      return {
        redirect: {
          destination: SIGN_IN_PAGE,
        },
      }
    }

    const providers = await getProviders()

    if (providers?.oidc) {
      return {
        props: {
          isAuthenticated: Boolean(session),
          csrfToken: await getCsrfToken({ req: context.req }),
          callbackUrl: `/api/signout?idTokenHint=${idTokenHint}`,
        },
      }
    }

    throw new Error('Could not retrieve auth providers')
  } catch (error) {
    logger.info(error)
    return {
      redirect: {
        destination: '/500',
      },
    }
  }
}
