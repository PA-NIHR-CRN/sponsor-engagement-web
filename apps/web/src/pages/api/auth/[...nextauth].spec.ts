import { faker } from '@faker-js/faker'
import type { refreshTokenResponseSchema } from '@nihr-ui/auth'
import { authService } from '@nihr-ui/auth'
import { logger } from '@nihr-ui/logger'
import type { Prisma, UserOrganisation, UserRole } from 'database'
import type { Account, Session } from 'next-auth'
import type { AdapterUser } from 'next-auth/adapters'
import type { JWT } from 'next-auth/jwt'
import type { Provider } from 'next-auth/providers'
import { Mock } from 'ts-mockery'
import { type z } from 'zod'

import { prismaMock } from '@/__mocks__/prisma'
import { Roles } from '@/constants/auth'

import { authOptions } from './[...nextauth]'

jest.mock('@nihr-ui/auth')
jest.mock('@nihr-ui/logger')
jest.useFakeTimers().setSystemTime(new Date('2023-01-01'))

beforeEach(() => {
  jest.clearAllMocks()
})

describe('Authentication configuration options', () => {
  test('authOptions object is defined', () => {
    expect(authOptions).toMatchSnapshot()
  })
})

describe('Custom OAuth provider is compatible with next-auth', () => {
  test('provider object is defined', () => {
    expect(authOptions.providers[0]).toEqual<Provider>({
      id: 'oidc',
      name: 'OIDC',
      type: 'oauth',
      wellKnown: 'mockedWellKnownUrl',
      clientId: 'mockedClientId',
      clientSecret: 'mockedClientSecret',
      allowDangerousEmailAccountLinking: true,
      authorization: { params: { scope: 'openid email profile' } },
      idToken: true,
      checks: ['pkce', 'state'],
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- valid any
      profile: expect.any(Function),
    })
  })
})

describe('JWT callback', () => {
  const accountMock = Mock.of<Account>({
    accessToken: faker.string.uuid(),
    expires_at: Number(faker.date.future()),
    refresh_token: faker.string.uuid(),
    id_token: faker.string.uuid(),
  })

  const userMock = Mock.of<AdapterUser>({
    email: faker.internet.email(),
  })

  const tokenMock = Mock.of<JWT>()

  test('returns a new token with account information on initial sign in', async () => {
    const account = accountMock
    const user = userMock
    const token = tokenMock

    const response = await authOptions.callbacks?.jwt?.({ account, user, token })

    expect(response).toEqual<JWT>({
      accessToken: accountMock.access_token,
      accessTokenExpires: accountMock.expires_at * 1000,
      refreshToken: accountMock.refresh_token,
      idToken: accountMock.id_token,
      user: userMock,
    })
  })

  test('returns the existing token when it has not expired yet', async () => {
    const account = null
    const user = userMock
    const accessTokenExpires = Number(faker.date.future())
    const token = Mock.of<JWT>({
      accessToken: accountMock.access_token,
      accessTokenExpires,
      refreshToken: accountMock.refresh_token,
      idToken: accountMock.id_token,
      user: userMock,
    })

    const response = await authOptions.callbacks?.jwt?.({ account, user, token })

    expect(response).toEqual<JWT>(token)
  })

  test('refreshes the token when it has expired', async () => {
    const account = null
    const user = userMock
    const accessTokenExpires = Number(faker.date.past())
    const token = Mock.of<JWT>({
      accessToken: accountMock.access_token,
      accessTokenExpires,
      refreshToken: accountMock.refresh_token,
      idToken: accountMock.id_token,
      user: userMock,
    })

    const refreshTokenResponseMock = Mock.of<z.infer<typeof refreshTokenResponseSchema>>({
      access_token: faker.string.uuid(),
      expires_in: Number(faker.date.future()),
      id_token: faker.string.uuid(),
      refresh_token: faker.string.uuid(),
    })

    jest.mocked(authService.refreshToken).mockResolvedValueOnce({
      success: true,
      data: refreshTokenResponseMock,
    })

    const response = await authOptions.callbacks?.jwt?.({ account, user, token })

    expect(response).toEqual<JWT>({
      ...token,
      accessToken: refreshTokenResponseMock.access_token,
      accessTokenExpires: Date.now() + refreshTokenResponseMock.expires_in * 1000,
      idToken: refreshTokenResponseMock.id_token,
      refreshToken: refreshTokenResponseMock.refresh_token,
    })
  })

  test('refresh token endpoint returns a 400 error due to refresh token expiry', async () => {
    const account = null
    const user = userMock
    const accessTokenExpires = Number(faker.date.past())
    const token = Mock.of<JWT>({
      accessToken: accountMock.access_token,
      accessTokenExpires,
      refreshToken: accountMock.refresh_token,
      idToken: accountMock.id_token,
      user: userMock,
    })

    jest.mocked(authService.refreshToken).mockRejectedValueOnce({
      isAxiosError: true,
      response: {
        data: {
          error: 'invalid_grant',
        },
      },
    })

    const response = await authOptions.callbacks?.jwt?.({ account, user, token })

    expect(logger.info).toHaveBeenCalledWith('jwt callback - refresh token expired - redirecting to login')

    expect(response).toEqual<JWT>({
      ...token,
      error: 'RefreshAccessTokenError',
    })
  })

  test('refresh token endpoint returns an unexpected error', async () => {
    const account = null
    const user = userMock
    const accessTokenExpires = Number(faker.date.past())
    const token = Mock.of<JWT>({
      accessToken: accountMock.access_token,
      accessTokenExpires,
      refreshToken: accountMock.refresh_token,
      idToken: accountMock.id_token,
      user: userMock,
    })

    jest.mocked(authService.refreshToken).mockRejectedValueOnce({
      isAxiosError: true,
      response: {
        data: {},
      },
    })

    const response = await authOptions.callbacks?.jwt?.({ account, user, token })

    expect(logger.error).toHaveBeenCalledWith({ isAxiosError: true, response: { data: {} } })

    expect(response).toEqual<JWT>({
      ...token,
      error: 'RefreshAccessTokenError',
    })
  })
})

describe('Session callback', () => {
  test('retrieves the roles for the authenticated user and forwards it to the returned session object', async () => {
    const providerUserId = 'provider-user-id'
    const localUserId = 1
    const email = faker.internet.email()
    const name = faker.person.fullName()

    const session = Mock.of<Session>({ user: { name, email } })
    const user = Mock.of<AdapterUser>()
    const token = Mock.of<JWT>({ user: { id: providerUserId, email } })

    prismaMock.user.findFirstOrThrow.mockResolvedValueOnce(
      Mock.of<Prisma.UserGetPayload<undefined>>({
        id: localUserId,
      })
    )

    prismaMock.userRole.findMany.mockResolvedValueOnce([
      Mock.of<UserRole>({
        userId: Number(providerUserId),
        roleId: Roles.SponsorContact,
      }),
      Mock.of<UserRole>({
        userId: Number(providerUserId),
        roleId: Roles.ContactManager,
      }),
    ])

    const mockUserOrganisation = Mock.of<UserOrganisation>({ organisationId: 123 })

    prismaMock.userOrganisation.findMany.mockResolvedValueOnce([mockUserOrganisation])

    const updatedSession = await authOptions.callbacks?.session?.({
      session,
      user,
      token,
      newSession: false,
      trigger: 'update',
    })

    expect(updatedSession).toEqual<Session>(
      Mock.of<Session>({
        error: undefined,
        idleTimeout: 600,
        user: {
          id: localUserId,
          name,
          email,
          roles: [Roles.SponsorContact, Roles.ContactManager],
          organisations: [mockUserOrganisation],
          groups: [],
        },
      })
    )
  })

  test('handles errors when fetching user', async () => {
    const user = Mock.of<AdapterUser>()
    const token = Mock.of<JWT>({ user: { id: faker.string.alphanumeric(5), email: faker.internet.email() } })
    const session = Mock.of<Session>({
      user: {},
    })

    prismaMock.user.findFirstOrThrow.mockRejectedValueOnce('user not found')

    const updatedSession = await authOptions.callbacks?.session?.({
      session,
      user,
      token,
      newSession: false,
      trigger: 'update',
    })

    expect(logger.error).toHaveBeenCalledWith('NextAuth session callback exception for %s', token.user.email)
    expect(logger.error).toHaveBeenCalledWith('user not found')

    expect(updatedSession).toEqual<Session>(
      Mock.of<Session>({
        user: {},
      })
    )
  })

  test('exits early if no user object exists within the session', async () => {
    const user = Mock.of<AdapterUser>()
    const token = Mock.of<JWT>()
    const session = Mock.of<Session>({
      user: null,
    })

    const updatedSession = await authOptions.callbacks?.session?.({
      session,
      user,
      token,
      newSession: false,
      trigger: 'update',
    })

    expect(prismaMock.userRole.findMany).not.toHaveBeenCalled()

    expect(updatedSession).toEqual<Session>(session)
  })
})

describe('signIn callback', () => {
  const accountMock = Mock.of<Account>({
    accessToken: faker.string.uuid(),
    expires_at: Number(faker.date.future()),
    refresh_token: faker.string.uuid(),
    id_token: faker.string.uuid(),
  })

  const userMock = Mock.of<AdapterUser>({
    email: faker.internet.email(),
  })

  test('records the last login time for the user', async () => {
    const account = accountMock
    const user = userMock

    const response = await authOptions.callbacks?.signIn?.({ account, user })

    expect(prismaMock.user.update).toHaveBeenCalledWith({
      where: {
        email: user.email,
      },
      data: {
        lastLogin: new Date('2023-01-01'),
      },
    })

    expect(response).toBe(true)
  })

  test('handles failures when updating user record', async () => {
    const account = accountMock
    const user = userMock

    prismaMock.user.update.mockRejectedValueOnce('test error')

    const response = await authOptions.callbacks?.signIn?.({ account, user })

    expect(logger.error).toHaveBeenCalledWith('NextAuth signIn callback exception for %s', userMock.email)
    expect(logger.error).toHaveBeenCalledWith('test error')

    expect(response).toBe(true)
  })
})
