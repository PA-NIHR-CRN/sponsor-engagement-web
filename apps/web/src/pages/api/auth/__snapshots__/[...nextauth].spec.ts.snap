// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication configuration options authOptions object is defined 1`] = `
{
  "callbacks": {
    "jwt": [Function],
    "redirect": [Function],
    "session": [Function],
  },
  "debug": true,
  "providers": [
    {
      "allowDangerousEmailAccountLinking": true,
      "authorization": {
        "params": {
          "scope": "openid email profile",
        },
      },
      "checks": [
        "pkce",
        "state",
      ],
      "clientId": "mockedClientId",
      "clientSecret": "mockedClientSecret",
      "id": "oidc",
      "idToken": true,
      "name": "OIDC",
      "profile": [Function],
      "type": "oauth",
      "wellKnown": "mockedWellKnownUrl",
    },
  ],
  "session": {
    "maxAge": 2592000,
    "strategy": "jwt",
  },
}
`;
